server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.features;
+
+import org.elasticsearch.Version;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * Service responsible for registering features this node should publish that it has.
+ */
+public class FeatureService {
+
+    private static final Map<Version, Set<String>> HISTORICAL_FEATURES = Map.ofEntries();
+
+    private static final NavigableMap<Version, Set<String>> AGGREGATED_FEATURES;
+
+    static {
+        NavigableMap<Version, Set<String>> consolidated = new TreeMap<>(HISTORICAL_FEATURES);
+        int thisEra = Version.CURRENT.major;
+
+        // add in all features from previous versions
+        Set<String> aggregatedFeatures = new HashSet<>();
+        for (Map.Entry<Version, Set<String>> versions : consolidated.entrySet()) {
+            if (versions.getKey().major < thisEra - 1) continue;    // don't include, its past the valid eras
+
+            // check the sizes to ensure we haven't got dups
+            int sizeBefore = aggregatedFeatures.size();
+            aggregatedFeatures.addAll(versions.getValue());
+            if (aggregatedFeatures.size() < sizeBefore + versions.getValue().size()) {
```

Comment:
I think addAll on a Set may already remove duplicates, but need to check

Reply:
Oh I see, you are leveraging that to actually detect this condition. 
Why do we care?

Reply:
To check if there are any duplicates in the historical versions list above (which is incorrect)